### ISSUE TITLE
Mypy does not detect automatic init generation for fieldless dataclass inheriting from abstract base

### ISSUE THREAD

**Issue by Nuno-Mota** [python/mypy#16613]
**Bug Report**
It appears that mypy ignores that the `__init__` of dataclasses is automatically generated even when they have no fields.

**To Reproduce**

Consider the following `abstract_dataclass.py` MWE:
```python
from abc import ABC, abstractmethod
from dataclasses import dataclass

class Foo(ABC):
	@abstractmethod
	def __init__(self) -> None: ...

	def __post_init__(self) -> None:
		print(f"Created '{self.__class__.__name__}' instance.")

@dataclass
class Bar(Foo): ...

@dataclass
class Baz(Foo):
	a: int

Bar()    # error: Cannot instantiate abstract class "Bar" with abstract attribute "__init__" [abstract]
Baz(a=3) # No error.
```

Running it, however:
```
python abstract_dataclass.py
Created 'Bar' instance.
Created 'Baz' instance.
```

**Expected Behavior**

No error should be identified.

**Your Environment**

- Mypy version used: 1.3.0
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.10.8

**Comment by Yikai-Liao** [comment]
There are so many inconsistencies among     different Python type checkers that I feel like I'm looking at different c++ compilers😭
